### PR TITLE
router: Store `recognize` outside of lock

### PR DESCRIPTION
**Pull request by olix0r**
The router stores its cache and `Recognize` implementation within a `Mutex`,
but there is no need for the recognizer to be locked.

This change creates a new `Cache` type that is locked independently of
`Recognize`. In order to accomplish this, `Recognize::bind_service` has
been changed to take an immutable reference to its `self`.

The (unused) `Single` type has been removed because it relied on
`bind_service` being mutable.

### CODE REVIEW

File: proxy/router/src/lib.rs [hunk 1]
```diff
@@ -23,7 +25,7 @@ where T: Recognize,
 /// `recognize()` method is used to determine the key for a given request. This key is
 /// used to look up a route in a cache (i.e. in `Router`), or can be passed to
 /// `bind_service` to instantiate the identified route.
-pub trait Recognize {
+pub trait Recognize: Clone {
```
Comment (seanmonstar) [inline proxy/router/src/lib.rs]: One downside perhaps not immediately obvious is that `Clone` isn't object-safe, so this prevents a `Router` from ever using trait objects. We may not need them in the proxy, but it'd probably be needed in a more generic `tower-router`.

Comment (carllerche) [reply proxy/router/src/lib.rs]: It doesn't fully prevent it. There are ways around it w/ additional `clone_box` `clone_arc` etc... fns + a special trait obj `Recognize` wrapper....

Comment (olix0r) [reply proxy/router/src/lib.rs]: Is it the same/better/worse if we move the Clone requirement onto Router (i.e. `T: Recognize + Clone`)?

Comment (seanmonstar) [reply proxy/router/src/lib.rs]: It's pretty much the same.

Here's a thought: does `Recognize::bind_service` need to be `&mut self`? It feels very similar to `NewService::new_service`, which takes `&self`. If it didn't need to be mutable, then we could just put the `Recognize` inside the `Arc`, but outside the `Mutex`.

Comment (olix0r) [reply proxy/router/src/lib.rs]: @seanmonstar within the scope of conduit, that works just fine

File: proxy/router/src/lib.rs [hunk 2]
```diff
@@ -73,8 +71,13 @@ where T: Recognize,
 struct Inner<T>
 where T: Recognize,
 {
-    routes: IndexMap<T::Key, T::Service>,
     recognize: T,
+    cache: Mutex<Cache<T::Key, T::Service>>,
```
Comment (carllerche) [inline proxy/router/src/lib.rs]: If we are nit picking, why not move `Mutex` into `Cache`, then the lock is fully encapsulated and you could remove it from there later.

This is not a critical change, just a thought.

Comment (seanmonstar) [reply proxy/router/src/lib.rs]: Then you'd need yet another struct, since there's multiple fields in `Cache` that need to be accessed behind the lock..

Comment (olix0r) [reply proxy/router/src/lib.rs]: really, only the map needs to be locked. i'll make this change with the followup Cache change https://github.com/runconduit/conduit/pull/920

**Pull request merged**